extensions/mssql/src/features.ts

```diff
@@ -30,6 +32,68 @@ export class TelemetryFeature implements StaticFeature {
 	}
 }
 
+export class NativeCredentialsFeature extends SqlOpsFeature<any>  {
+
+	private keychain: Keychain;
+
+	private static readonly messagesTypes: RPCMessageType[] = [
+		DeleteCredentialRequest.type,
+		SaveCredentialRequest.type,
+		ReadCredentialRequest.type
+	];
+
+	fillClientCapabilities(capabilities: ClientCapabilities): void {
+		Utils.ensure(Utils.ensure(capabilities, 'credentials')!, 'credentials')!.dynamicRegistration = true;
+	}
+
+	initialize(capabilities: ServerCapabilities): void {
+		this.register(this.messages, {
+			id: UUID.generateUuid(),
+			registerOptions: undefined
+		});
+	}
+
+	constructor(_client: SqlOpsDataClient) {
+		super(_client, NativeCredentialsFeature.messagesTypes);
+		this.keychain = new Keychain();
+	}
+
+	protected registerProvider(options: any): Disposable {
+
+		let readCredential = (credentialId: string): Thenable<azdata.Credential> => {
+			return this.keychain.getPassword(credentialId).then((password) => {
+				if (password) {
+					const credential: azdata.Credential = {
+						credentialId: credentialId,
+						password: password
+					};
+					return credential;
+				}
+				return undefined;
+			});
+		};
+
+		let saveCredential = (credentialId: string, password: string): Thenable<boolean> => {
+			return this.keychain.setPassword(credentialId, password).then(() => {
+				return this.keychain.getPassword(credentialId).then((savedPassword) => {
+					return savedPassword === password;
+				});
+			});
+		};
+
+		let deleteCredential = (credentialId: string): Thenable<boolean> => {
+			return this.keychain.deletePassword(credentialId).then((result) => { return result; });
```

Comment:
do we need the `.then` block here? or can we return the promise coming back from `deletePassword` directly?